### PR TITLE
feat(libs): add barcode widget

### PR DESCRIPTION
**Pull request by FASTSHIFT**
### Description of the feature or fix

Ported from: https://github.com/fhunleth/code128

image

### Checkpoints
- [x] Run `code-format.py` from the scripts folder. [astyle](http://astyle.sourceforge.net/install.html) needs to be installed.
- [x] Update the [Documentation](https://github.com/lvgl/lvgl/tree/master/docs) if needed
- [x] Add [Examples](https://github.com/lvgl/lvgl/tree/master/examples) if relevant. 
- [x] Add [Tests](https://github.com/lvgl/lvgl/blob/master/tests/README.md) if applicable.
- [x] If you added new options to `lv_conf_template.h` run [lv_conf_internal_gen.py](https://github.com/lvgl/lvgl/blob/release/v8.3/scripts/lv_conf_internal_gen.py) and update [Kconfig](https://github.com/lvgl/lvgl/blob/release/v8.3/Kconfig).
 
Be sure the following conventions are followed:
- [x] Follow the [Styling guide](https://github.com/lvgl/lvgl/blob/master/docs/CODING_STYLE.md)
- [x] Prefer `enum`s instead of macros. If inevitable to use `define`s export them with `LV_EXPORT_CONST_INT(defined_value)` right after the `define`.
- [x] In function arguments prefer `type name[]` declaration for array parameters instead of `type * name`
- [x] Use typed pointers instead of `void *` pointers
- [x] Do not `malloc` into a static or global variables. Instead declare the variable in `LV_ITERATE_ROOTS` list in [`lv_gc.h`](https://github.com/lvgl/lvgl/blob/master/src/misc/lv_gc.h) and mark the variable with `GC_ROOT(variable)` when it's used.
- [x] Widget constructor must follow the `lv_<widget_name>_create(lv_obj_t * parent)` pattern.
- [x] Widget members function must start with `lv_<modul_name>` and should receive `lv_obj_t *` as first argument which is a pointer to widget object itself.  
- [x] `struct`s should be used via an API and not modified directly via their elements.
- [x] `struct` APIs should follow the widgets' conventions. That is to receive a pointer to the `struct` as the first argument, and the prefix of the `struct` name should be used as the prefix of the function name too (e.g.  `lv_disp_set_default(lv_disp_t * disp)`)
- [x] Functions and `struct`s which are not part of the public API must begin with underscore in order to mark them as "private".
- [x] Arguments must be named in H files too.
- [x] To register and use callbacks one of the followings needs to be followed: 
  - For both the registration function and the callback pass a pointer to a `struct` as the first argument. The `struct` must contain `void * user_data` field.
  - The last argument of the registration function must be `void * user_data` and the same `user_data` needs to be passed as the last argument of the callback.
  - Callback types not following these conventions should end with `xcb_t`.

### CODE REVIEW

File: examples/libs/barcode/lv_example_barcode_1.py [hunk 1]
```diff
@@ -0,0 +1,21 @@
+#!/opt/bin/lv_micropython -i
+import lvgl as lv
+import display_driver
+
+bg_color = lv.palette_lighten(lv.PALETTE.LIGHT_BLUE, 5)
+fg_color = lv.palette_darken(lv.PALETTE.BLUE, 4)
+
+barcode = lv.barcodecode(lv.scr_act())
```
Comment (kisvegabor) [inline examples/libs/barcode/lv_example_barcode_1.py]:
```suggestion
barcode = lv.barcode(lv.scr_act())
```

**Review by kisvegabor — COMMENTED** [review]
Very nice! Do you have a our test environment set locally? If you could you add simple [rendering tests](https://github.com/lvgl/lvgl/blob/master/tests/README.md#asserts) with a few bar codes with different colors?

**Comment by kisvegabor** [comment]
@amirgon Do you have any idea why the MP CI fails?

**Comment by FASTSHIFT** [comment]
@kisvegabor 
`LV_USE_BARCODE` is 0 by default. How to automatically enable it to test and generate reference images?

**Comment by kisvegabor** [comment]
Please enable it here:
https://github.com/lvgl/lvgl/blob/b01229e451289ae9c64c51897bfb70045227080b/tests/CMakeLists.txt#L249

**Comment by amirgon** [comment]
> @amirgon Do you have any idea why the MP CI fails?

Yes. `lv_conf.h` didn't enable `LV_USE_BARCODE` therefore "barcode" widget wasn't built for the Micropython CI.
I've updated lv_micropython to enable it, please try to re-run the test.

In the longer term, it might make sense to move Micropython's lv_conf.h to lvgl repo, such that an LVGL PR could update it. Otherwise additional commits to [lv_binding_micropython](https://github.com/lvgl/lv_binding_micropython/commit/c544c51c97b38907709a0cd7ea7ca8ab96410cbb) and [lv_micropython](https://github.com/lvgl/lv_micropython/commit/b8dde900fbe08bde418b7d7c1c51f6c060c355ab) are required.

**Comment by embeddedt** [comment]
> In the longer term, it might make sense to move Micropython's lv_conf.h to lvgl repo, such that an LVGL PR could update it. 

I agree; we seem to have this problem regularly.

**Comment by kisvegabor** [comment]
> In the longer term, it might make sense to move Micropython's lv_conf.h to lvgl repo, such that an LVGL PR could update it. Otherwise additional commits to [lv_binding_micropython](https://github.com/lvgl/lv_binding_micropython/commit/c544c51c97b38907709a0cd7ea7ca8ab96410cbb) and [lv_micropython](https://github.com/lvgl/lv_micropython/commit/b8dde900fbe08bde418b7d7c1c51f6c060c355ab) are required.

The idea is that the content of `lvgl` shouldn't be modified by the user. If we added an `lv_conf_mp.h` here the user needed to edit it directly. What about adding a CMake or Makefile for the MP tests which defines the correct settings for MP tests and `MP's lv_conf.h` wouldn't be not used for the tests?

**Comment by amirgon** [comment]
> The idea is that the content of `lvgl` shouldn't be modified by the user. If we added an `lv_conf_mp.h`

But LVGL should provide some defaults, no?  
If someone adds a new widget, how can we test it if it's not enabled by default?

> What about adding a CMake or Makefile for the MP tests which defines the correct settings for MP tests and `MP's lv_conf.h` wouldn't be not used for the tests?

Do you mean add some CMake or Makefile in LVGL repo? Today all make logic is in the lv_micropython and lv_binding_micropython repos. Does it make sense to add Micropython make logic to LVGL?

How can we not use MP's lv_conf.h for the tests? Where will we take all the configuration from? How do we update the configuration without duplicating it across repos?

**Comment by kisvegabor** [comment]
> If someone adds a new widget, how can we test it if it's not enabled by default?

From [`lv_conf_template.h`](https://github.com/lvgl/lvgl/blob/master/lv_conf_template.h) we generate [`lv_conf_internal.h`](https://github.com/lvgl/lvgl/blob/master/src/lv_conf_internal.h). So the defaults are in `lv_conf_template.h` and if no one set those defines `lv_conf_internal.h` will set the default value.

> Does it make sense to add Micropython make logic to LVGL?

As we have [CI test for MP](https://github.com/lvgl/lvgl/blob/master/.github/workflows/build_micropython.yml) also in the `lvgl` repo an `mk` or `h` file in the [`tests`](https://github.com/lvgl/lvgl/tree/master/tests) folder can be ok too.

> How can we not use MP's lv_conf.h for the tests? Where will we take all the configuration from? How do we update the configuration without duplicating it across repos?

With `-DLV_CONF_PATH=".../my_custom_conf.h"` we can use a custom `lv_conf.h` file. We already maintain a list of test configs  in [CMakeLists.txt](https://github.com/lvgl/lvgl/blob/master/tests/CMakeLists.txt) but it's CMake, not make. So we can't reuse them. :slightly_frowning_face:

**Comment by amirgon** [comment]
> From [`lv_conf_template.h`](https://github.com/lvgl/lvgl/blob/master/lv_conf_template.h) we generate [`lv_conf_internal.h`](https://github.com/lvgl/lvgl/blob/master/src/lv_conf_internal.h). So the defaults are in `lv_conf_template.h` and if no one set those defines `lv_conf_internal.h` will set the default value.

But the default values are false... Here on this PR for example `lv_conf_template.h`:
```c
/*Barcode code library*/
#define LV_USE_BARCODE 0
```
We need it to be 1 for the tests to pass, so `lv_conf_template.h` doesn't help here.

> As we have [CI test for MP](https://github.com/lvgl/lvgl/blob/master/.github/workflows/build_micropython.yml) also in the `lvgl` repo an `mk` or `h` file in the [`tests`](https://github.com/lvgl/lvgl/tree/master/tests) folder can be ok too.

I still don't see the point in adding Micropython specific build scripts in LVGL.  
Wouldn't it make more sense to keep all build scripts in one place, and only pass them parameters?

> With `-DLV_CONF_PATH=".../my_custom_conf.h"` we can use a custom `lv_conf.h` file. We already maintain a list of test configs in [CMakeLists.txt](https://github.com/lvgl/lvgl/blob/master/tests/CMakeLists.txt) but it's CMake, not make. So we can't reuse them.

Why do you keep these test defaults in CMake and not in a simple plain H file?  
If you had something like `lv_conf_test.h` file which configures LVGL for tests, we could have used it directly for both Micropython and LVGL tests without code duplication and without any specialized make scripts.

**Comment by kisvegabor** [comment]
Let me reiterate: instead of having CMake configs we could have an `lv_test_conf.h` file, in which based on the test settings (`-DTEST_TYPE`) we can `#if #endif` some parts. Actually we already have [`lv_test_conf.h`](https://github.com/lvgl/lvgl/blob/master/tests/src/lv_test_conf.h) but it's very simple now.  

The gist is that we use these `lv_test_conf.h` only for the tests (the values defined in `lv_test_conf.h` overwrite the default values).

**Comment by amirgon** [comment]
> Let me reiterate: instead of having CMake configs we could have an `lv_test_conf.h` file, in which based on the test settings (`-DTEST_TYPE`) we can `#if #endif` some parts. Actually we already have [`lv_test_conf.h`](https://github.com/lvgl/lvgl/blob/master/tests/src/lv_test_conf.h) but it's very simple now.

Ok. Having a single `lv_test_conf.h` for both C and MP would make sense, to prevent code duplication.
It would probably need to undef existing defines, since most configurations in `lv_conf.h` are not conditioned.

**Comment by kisvegabor** [comment]
> It would probably need to undef existing defines, since most configurations in lv_conf.h are not conditioned.

We can use only one `lv_conf.h` at a time. For tests it'd be `lv_test_conf.h` so it doesn't interfere with the normal `lv_conf.h`. 

So in summary:
- There is an `lv_test_conf.h` in `lvgl/tests`
- This `h` file has the configs for various test settings include MP tests
- MP test builds shall add `-DLV_CONF_PATH="lvgl/tests/lv_test_conf.h"`

What do you think?

**Comment by amirgon** [comment]
> We can use only one `lv_conf.h` at a time. For tests it'd be `lv_test_conf.h` so it doesn't interfere with the normal `lv_conf.h`.

In case of Micropython we already have [lv_conf.h](https://github.com/lvgl/lv_binding_micropython/blob/master/lv_conf.h) which is custom made for the Micropython bindings. All we really need is to enable some widgets so we could test them.  
That's why I suggested that `lv_test_conf.h` would **override** it, not replace it.

Does it make any sense to maintain two copies of almost identical `lv_conf.h` files?  
We could easily avoid this code duplication in several ways:
- We could override macros, by undefining and redefining them to enable widgets for tests
- We could enable a widget in ifdef in lv_conf.h, so it takes effect only if it wasn't enabled/disabled already. We already use this technique today to allow [setting color depths from outside](https://github.com/lvgl/lv_binding_micropython/blob/c544c51c97b38907709a0cd7ea7ca8ab96410cbb/lv_conf.h#L27-L29) for example.

**Comment by kisvegabor** [comment]
> In case of Micropython we already have [lv_conf.h](https://github.com/lvgl/lv_binding_micropython/blob/master/lv_conf.h) which is custom made for the Micropython bindings. All we really need is to enable some widgets so we could test them.
That's why I suggested that lv_test_conf.h would override it, not replace it.

I see your point. As LVGL has no config to handle multiple `lv_conf` files we could do something like this is `lv_test_conf.h`:
```c
#ifdef MICROPYTHON
  #include "../../../lv_conf.h"  //from lv_binding_micropython

  #undef LV_USE_BTN
  #undef LV_USE_LABEL
  #undef LV_USE_BARCODE
  ...
#endif

#if TEST_TYPE == 1
  #include "test_type_1_conf.h"   
#endif

#if TEST_TYPE == 2
  #include "test_type_2_conf.h"   
#endif
```  

This way all the extra work we need is to `undef` all the option we will set later in  `test_type_*_conf.h`.

What do you think?

**Comment by amirgon** [comment]
> I see your point. As LVGL has no config to handle multiple `lv_conf` files we could do something like this is `lv_test_conf.h`:
> ...
> This way all the extra work we need is to `undef` all the option we will set later in `test_type_*_conf.h`.
> 
> What do you think?

In principle this looks good to me.

But I'm still not sure I understand the details:
- What kind of "test types" would we have? How would `test_type_1_conf.h` look like for example?
- I assume that these `test_type_X_conf.h` would only enable widgets, is that right? Or would they affect other configurations as well?
- Since these "test types" are set on compile time, only one of them would be relevant for Micropython right? Micropython has only a single compile time configuration. How would `TEST_TYPE` be configured when we build Micropython for tests?
- If there's only a single `TEST_TYPE` for Micropython testing, why do we need to undefine these macros globally and not in the context of that specific `TEST_TYPE`?

**Comment by kisvegabor** [comment]
> What kind of "test types" would we have? How would test_type_1_conf.h look like for example?

These:
image




> I assume that these test_type_X_conf.h would only enable widgets, is that right? Or would they affect other configurations as well?

Enable widgets and other features. E.g. `LV_USE_FONT_SUBPX`

> Since these "test types" are set on compile time, only one of them would be relevant for Micropython right? Micropython has only a single compile time configuration. How would TEST_TYPE be configured when we build Micropython for tests?
If there's only a single TEST_TYPE for Micropython testing, why do we need to undefine these macros globally and not in the context of that specific TEST_TYPE?

You are right. Probably it should be better:
```c
#ifdef MICROPYTHON
  #include "../../../lv_conf.h"  //from lv_binding_micropython
#endif

#if TEST_TYPE == 1
  #include "test_type_1_conf.h"   
#endif

#if TEST_TYPE == 2
  #include "test_type_2_conf.h"   
#endif
```

`#undef` is not required as each test type has it' own `lv_conf.h` like H file.

**Comment by amirgon** [comment]
> `#undef` is not required as each test type has it' own `lv_conf.h` like H file.

Are you sure?  
What if both the binding's `lv_conf.h` and the test's `lv_conf.h` define/undefine the same macro?

**Comment by kisvegabor** [comment]
In MP test we use only the bindings `lv_conf.h`. It will serve as the test config. Just like now. 

But I think we lost the purpose (or maybe just me).  The original goal was to enable the new widgets only on one place. And it should be the `test_type_X_conf.h` files. So what about this:
```c
//Platform specific configs
#ifdef MICROPYTHON
  #include "../../../lv_conf.h"     //from lv_binding_micropython, memory config + enable widgets
#endif

#ifdef UNIX
  #include "unix_test_conf.h"        // Just memory and other basic settings 
#endif


#if TEST_TYPE == 1
  #include "test_disable_all.h"   //#undef and disable all features to have a clean start. Can be included form test_type_1_conf.h too
  #include "test_type_1_conf.h"   //#undef what it will overwrite
#endif

#if TEST_TYPE == 2
  #include "test_disable_all.h"   //#undef and disable all features to have a clean start. Can be included form test_type_2_conf.h too
  #include "test_type_2_conf.h"   
#endif
```

The MP test should `-DTEST_TYPE=5` (let's say in 5 everything is enabled).

**Comment by amirgon** [comment]
> In MP test we use only the bindings `lv_conf.h`. It will serve as the test config. Just like now.

The problem is that when a new widget is added and the define is still missing in `lv_conf.h`. That was the original reason MP test failed on this PR.

> The MP test should `-DTEST_TYPE=5` (let's say in 5 everything is enabled).

Yes assuming it would also ` #include "test_disable_all.h"` first.
As a test that enables everything, how about not defining `TEST_TYPE` at all? (then check it with `#ifndef`)

**Comment by kisvegabor** [comment]
> Yes assuming it would also #include "test_disable_all.h" first.

Yes, we'd keep only the memory and other platform specific settings.

> As a test that enables everything, how about not defining TEST_TYPE at all? (then check it with #ifndef)

It could also work. E.g.
```c
#if !defined(TEST_TYPE) || TEST_TYPE == 5
  #include "test_disable_all.h" 
  #include "test_type_5_conf.h"  
#endif
```

**Comment by FASTSHIFT** [comment]
Having trouble with the MP test? Why do many PR tests fail.

**Comment by amirgon** [comment]
> Having trouble with the MP test? Why do many PR tests fail.

Only this test fails here:
```
micropython-dev: ../../lib/lv_bindings/lvgl/src/libs/tiny_ttf/stb_truetype_htcw.h:1781: stbtt_FindGlyphIndex: Assertion `0' failed.
```

Could you try to rebase your PR on latest LVGL? 

---

@kisvegabor Latest fails because you fixed only C and not Python examples:

https://github.com/lvgl/lvgl/commit/a218d7540945612d476f88b3b13d6a7a77b7e4de causes https://github.com/lvgl/lvgl/actions/runs/3502652826/jobs/5867181576

**Comment by kisvegabor** [comment]
> Latest fails because you fixed only C and not Python examples:

For some reason I haven't got email notification about the failing CI. Anyway, I've tried to fix it locally but got a strange error: 
```
$ ./run.sh 
Running the SDL lvgl version
Running ./../lvgl/examples/header.py ...
lvgl obj 
Running the SDL lvgl version
Running ./../lvgl/examples/event/lv_example_event_4.py ...
lvgl obj 
lvgl obj 
Traceback (most recent call last):
  File "./run_test.py", line 97, in run
  File "./run_test.py", line 47, in reraise
  File "lv_utils.py", line 123, in task_handler
  File "<string>", line 34, in event_cb
AttributeError: 'module' object has no attribute 'RADIUS_CIRCLE'
parallel: This job failed:
timeout 5m catchsegv ./../../../ports/unix/micropython-dev ./run_test.py ./../lvgl/examples/event/lv_example_event_4.py
```
I've confirmed that the built MicroPython recognizes `lv.RADIUS_CIRCLE`.

**Comment by amirgon** [comment]
> I've tried to fix it locally but got a strange error:
> ...
> I've confirmed that the built MicroPython recognizes lv.RADIUS_CIRCLE.

I checked out this PR and `lv_example_event_4.py` works fine.  
I'm not sure why you are seeing an error related to `lv_example_event_4.py`. Could you try to clean and rebuild?
Moreover, I don't see your error on any GitHub workflow, not on this PR and not on LVGL master.

What I do see, is an error with `lv_example_meter_3.py` and `lv_demo_widgets.py`:
```
AttributeError: 'meter' object has no attribute 'add_scale'
```
(https://github.com/lvgl/lvgl/actions/runs/3508673019/jobs/5877241151, https://github.com/lvgl/lvgl/actions/runs/3508136393/jobs/5876333063)

and this makes sense, because on https://github.com/lvgl/lvgl/commit/a218d7540945612d476f88b3b13d6a7a77b7e4de you removed `lv_meter_add_scale`

**Comment by kisvegabor** [comment]
> I'm not sure why you are seeing an error related to lv_example_event_4.py. Could you try to clean and rebuild?
Moreover, I don't see your error on any GitHub workflow, not on this PR and not on LVGL master.

It's a local issue for sure, but I don't know why. I've tried to clean and rebuild but didn't help.


> What I do see, is an error with lv_example_meter_3.py and lv_demo_widgets.py:

I've just fixed them. Now all tests pass.

**Comment by kisvegabor** [comment]
@amirgon 
I've added the new header based tests in [feat/test-configs-from-headers](https://github.com/lvgl/lvgl/tree/feat/test-configs-from-headers).

Finally, it became slightly different from what we have discussed because it made more sense in the practice. 
I've added MP's memory configs directly in [`lv_test_conf.h`](https://github.com/lvgl/lvgl/compare/lvgl:3940f53...lvgl:38a9f11#diff-e282720b51ca36e99b2b086a97641a2df42d4c3793485cfeb0e5975e06dbdd51R43-R61).  This way we don't need to `#undef` all the options defined by MP's `lv_conf.h`. I'm not sure I'm right, but I think all we need to do is making the tests on GitHub workflow to not use MP's `lv_conf.h` but [`lv_test_conf.h`](https://github.com/lvgl/lvgl/blob/feat/test-configs-from-headers/tests/src/lv_test_conf.h) with something like
`-DLV_CONF_PATH=${LVGL_TEST_DIR}/src/lv_test_conf.h`.

If a new widget/feature comes we only need to add it in [`lv_test_conf_full.h`](https://github.com/lvgl/lvgl/compare/lvgl:3940f53...lvgl:38a9f11#diff-05b9ed1a8384e4ba1f41d99452a63b732cac4e16faa882d2051a5ff7341978f7).

What do you think?

**Comment by kisvegabor** [comment]
Now it doesn't effect MicroPython so I merged the header based test to `master` in order to not block #3821.

**Comment by kisvegabor** [comment]
@embeddedt After this (test config to header files) change the ARM test on def heap fails with timeout on stress test:
image
https://github.com/lvgl/lvgl/actions/runs/3547541843/jobs/5957774176

Do you have any idea what could be the problem? The tests have 8MB memory.

**Comment by amirgon** [comment]
> I've added MP's memory configs directly in [`lv_test_conf.h`](https://github.com/lvgl/lvgl/compare/lvgl:3940f53...lvgl:38a9f11#diff-e282720b51ca36e99b2b086a97641a2df42d4c3793485cfeb0e5975e06dbdd51R43-R61)

@kisvegabor - That's a code duplication.  
If you do that, we'll always need to keep in mind that there are two different configuration files that need to be kept in sync, in two different repos. In my opinion this is confusing.  
For example, https://github.com/lvgl/lv_binding_micropython/pull/242 changes exacly this duplicated part.... Once it's merge, LVGL will break until we fix `lv_test_conf.h` on LVGL.
Another problem of this approach is dependencies. It requires specific versions of LVGL to match specific versions of lv_binding_micropython. Ideally we would want to be able to change LVGL submodule version in lv_binding_micropython and that it would just work. But if LVGL has Micropython configuration in it, that's another dependency that limits which versions of LVGL can work with which versions of lv_binding_micropython.   

So I would prefer to avoid this duplication and have only a single configuration.  
Since it's Micropython specific, I don't think it should be in LVGL repo.

There is another option, which is to define every configuration symbol only if it was not defined previously:
```c
#ifndef SYMBOL
#define SYNBOL ...
#endif
```
If all configuration macros everywhere were defined that way, it could have been very easy to override configuration just by including some header earlier, without the need to undef anything.

**Comment by kisvegabor** [comment]
What about creating an `lv_conf_mem.h` file in MP which could be reused by the tests and MP itself too?

> There is another option, which is to define every configuration symbol only if it was not defined previously:

If I understand correctly it'd apply to MP's `lv_conf.h` too. It'd make the already long file even longer and harder to understand. 

What is the correct `define` to detect MicroPython environment? I used `#ifdef MICROPYTHON` but it's just placeholder, I don't know if it's correct.

**Comment by amirgon** [comment]
> What about creating an `lv_conf_mem.h` file in MP which could be reused by the tests and MP itself too?

It's possible but doesn't solve the more general dependencies problem.  
If LVGL includes `lv_conf_mem.h` from Micropython then it assumes it exists and in specific path, which may or may not be true for a specific version of lv_binding_micropython.
Today we don't have assumptions in LVGL about Micropython, and this makes sense because LVGL is a library used by Micropython. It's better for a library to not have dependencies on the applications that use it.

> If I understand correctly it'd apply to MP's `lv_conf.h` too. It'd make the already long file even longer and harder to understand.

Yes it will be longer, but I don't think it will be harder to understand.  
In fact, on Micropython's `lv_conf.h` I've already started to do that [for some defines](https://github.com/lvgl/lv_binding_micropython/blob/c544c51c97b38907709a0cd7ea7ca8ab96410cbb/lv_conf.h#L27-L29).
It provides flexibility. For example, a user could provide any configuration parameter with the `-D` option on `CFLAGS`, even on the Makefile command line. [(`LV_CFLAGS="-DLV_COLOR_DEPTH=16"`)](https://github.com/lvgl/lv_micropython#esp32-port)

> What is the correct `define` to detect MicroPython environment? I used `#ifdef MICROPYTHON` but it's just placeholder, I don't know if it's correct.

I don't think we have such a define today.

**Comment by kisvegabor** [comment]
> Today we don't have assumptions in LVGL about Micropython, and this makes sense because LVGL is a library used by Micropython. It's better for a library to not have dependencies on the applications that use it.

What about a `LV_MP_TEST_CONF_PATH` define? Once we run MicroPython tests this dependency seems acceptable to me.

> I don't think we have such a define today.

We could use `LV_MP_TEST_CONF_PATH` for this purpose.

**Comment by amirgon** [comment]
> What about a `LV_MP_TEST_CONF_PATH` define? Once we run MicroPython tests this dependency seems acceptable to me.

What will be included in this `LV_MP_TEST_CONF_PATH`?  Only the macros that are specific to Micropython?  (gc, memory allocation)
What about `LV_EXPORT_CONST_INT` for example?

What you suggest can work, but requires me to split `lv_conf.h` in lv_bindings_micropython into two header files.  

In that case, it could be very useful for me to also change LVGL to have **two** header files instead of a single `lvgl/lv_conf_template.h`.   
That's because on every update I make to LVGL submodule in lv_bindings_micropython I always diff `lv_conf.h` vs. `lv_conf_template.h` to see what was changed or added there, then I update `lv_conf.h` accordingly. It would be harder for me to compare a single `lv_conf_template.h` to two header files in lv_bindings_micropython every time.

**Comment by kisvegabor** [comment]
> What will be included in this LV_MP_TEST_CONF_PATH? Only the macros that are specific to Micropython? (gc, memory allocation)
What about LV_EXPORT_CONST_INT for example?

Yes, `LV_EXPORT_CONST_INT` too. 

Realistically how often do you except these system level defines to change? If once in a year or so I think it's manageable to have  a separate `lv_conf_test.h` where the system level defines are copied from the normal `lv_conf.h`. So it's a duplication as `lv_conf_test.h` is not included  in `lv_conf.h`.  But it's flexible, simple, and doesn't require much change.

**Comment by FASTSHIFT** [comment]
@kisvegabor 
When is it expected to merge this PR? :)

**Comment by kisvegabor** [comment]
Let's merge it now. We can discuss the test updates independently.

**Pull request merged**